include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,501 @@
+        : m_polygon_projections(num_of_polygons)
+    {}
+
+    using coordinate_type = typename coordinate_type<PolyhedralSurface>::type;
+    using vec3 = boost::qvm::vec<coordinate_type, 3>;
+    using polygon_3d = typename boost::range_value<PolyhedralSurface>::type;
+    using ring_3d = typename boost::geometry::ring_type<polygon_3d>::type;
+    using point_3d = typename boost::geometry::point_type<ring_3d>::type;
+    using segment_3d = typename boost::geometry::model::segment<point_3d>;
+    using set_vec_from_point = typename strategy::transform::detail::matrix_transformer::
+        set_vec_from_point<point_3d, 0, 3>;
+
+    using point_2d = boost::geometry::model::point
+        <coordinate_type, 2, boost::geometry::cs::cartesian>;
+    using linestring_2d = typename boost::geometry::model::linestring<point_2d>;
+    using polygon_2d = typename std::conditional
+        <
+            boost::geometry::point_order<polygon_3d>::value == boost::geometry::counterclockwise,
+            typename boost::geometry::model::polygon<point_2d, true>,
+            typename boost::geometry::model::polygon<point_2d, false>
+        >::type;
+    using ring_2d = typename boost::geometry::ring_type<polygon_2d>::type;
+
+private:
+
+    // Tolerance for points considered colinear, coplanar, etc.
+    constexpr static coordinate_type tolerance = std::numeric_limits<coordinate_type>::epsilon();
+
+    enum class intersection_type
+    {
+        none,
+        valid,
+        invalid
+    };
+
+    struct polygon_projection
+    {
+        polygon_2d projected_polygon;
+        vec3 base;
+        vec3 v1;
+        vec3 v2;
+        vec3 normal;
+    };
+
+    template <int I, int J, typename Strategy>
+    static bool affinely_independent(point_3d const& p1, point_3d const& p2, point_3d const& p3,
+                                     Strategy const& strategy)
+    {
+        point_2d const p1_2{geometry::get<I>(p1), geometry::get<J>(p1)};
+        point_2d const p2_2{geometry::get<I>(p2), geometry::get<J>(p2)};
+        point_2d const p3_2{geometry::get<I>(p3), geometry::get<J>(p3)};
+        return strategy.side().apply(p1_2, p2_2, p3_2) != 0;
+    }
+
+    // Try to find 3 non-collinear points
+    template <typename Strategy>
+    static bool find_non_collinear_points(polygon_3d const& polygon, vec3& base, vec3& v1, vec3& v2,
+                                          Strategy const& strategy)
+    {
+        for (auto pit1 = points_begin(polygon); pit1 != points_end(polygon); ++pit1)
+        {
+            for (auto pit2 = pit1; pit2 != points_end(polygon); ++pit2)
+            {
+                for (auto pit3 = pit2; pit3 != points_end(polygon); ++pit3)
+                {
+                    if (affinely_independent<0, 1>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<0, 2>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<1, 2>(*pit1, *pit2, *pit3, strategy))
```

Comment:
Suggestion, add comment `One three-point-set is already sufficient, with such three independent points a plane can be constructed` (or similar).